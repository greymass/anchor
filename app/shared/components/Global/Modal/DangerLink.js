// @flow
import React, { Component } from 'react';
import { translate } from 'react-i18next';
import { Button, Header, Modal } from 'semantic-ui-react';

const { shell } = require('electron');

export class GlobalModalDangerLink extends Component<Props> {
  state = { open: false };

  isSafeish = (url) => (url && (url.startsWith('http:') || url.startsWith('https:') || url.startsWith('mailto:')))

  openLink() {
    const { link } = this.props;
    if (this.isSafeish(link)) {
      shell.openExternal(link);
    }
    this.setState({
      open: false
    });
  }

  openWarning() {
    this.setState({ open: true });
  }

  openLinkOrWarning() {
    const {
      settings
    } = this.props;

    if (settings.skipLinkModal) {
      this.openLink();
    } else {
      this.openWarning();
    }
  }

  close = () => this.setState({ open: false })

  render() {
    const {
      open
    } = this.state;
    const {
      content,
      link,
      settings,
      t
    } = this.props;
    let el = content || link;
    if (this.isSafeish(link)) {
      el = (
        <a
          onClick={this.openLinkOrWarning.bind(this)}
          onKeyPress={this.openLinkOrWarning.bind(this)}
          role="link"
          style={{ cursor: 'pointer' }}
          tabIndex={0}
        >
          {content || link}
        </a>
      );
    }
    return (
      <span>
        {el}
        {(!settings.skipLinkModal)
          ? (
            <Modal size="tiny" open={open} onClose={this.close}>
              <Modal.Header>
                {t('global_dangerlink_warning_title')}
              </Modal.Header>
              <Modal.Content>
                <p>
                  {t('global_dangerlink_warning_body_1')}
                </p>
                <p>
                  URL: {link}
                </p>
                <Header>
                  {t('global_dangerlink_warning_body_2')}
                </Header>
                <p>
                  {t('global_dangerlink_warning_body_3')}
                </p>
              </Modal.Content>
              <Modal.Actions>
                <Button
                  content={t('cancel')}
                  onClick={this.close}
                  negative
                />
                <Button
                  positive
                  onClick={this.openLink.bind(this)}
                  icon="external"
                  labelPosition="right"
                  content={t('confirm')}
                />
              </Modal.Actions>
            </Modal>
          ) : ''}
      </span>
    );
  }
}

export default translate('global')(GlobalModalDangerLink);
