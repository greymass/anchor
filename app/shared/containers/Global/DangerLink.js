// @flow
import React, { PureComponent } from 'react';
import { connect } from 'react-redux';

import DangerLink from '../../../shared/components/Global/Modal/DangerLink';

class GlobalBlockchainDangerLink extends PureComponent<Props> {
  render = () => (
    <DangerLink
      content={this.props.content}
      link={this.props.link}
      settings={this.props.settings}
    />
  );
}

function mapStateToProps(state) {
  return {
    settings: state.settings,
  };
}

export default connect(mapStateToProps)(GlobalBlockchainDangerLink);
