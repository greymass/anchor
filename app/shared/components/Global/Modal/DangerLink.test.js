import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { GlobalModalDangerLink } from './DangerLink';

const mockProps = {
  content: <React.Fragment />,
  link: 'https://link.com',
  settings: { skipLinkModal: true },
  t: text => text,
};

describe('GlobalModalDangerLink', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<GlobalModalDangerLink {...mockProps} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
