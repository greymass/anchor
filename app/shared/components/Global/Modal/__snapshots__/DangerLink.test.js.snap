// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalModalDangerLink renders correctly 1`] = `
<span>
  <a
    onClick={[Function]}
    onKeyPress={[Function]}
    role="link"
    style={
      Object {
        "cursor": "pointer",
      }
    }
    tabIndex={0}
  />
</span>
`;
